packages/ember-glimmer/tests/integration/components/append-test.js

```diff
@@ -0,0 +1,316 @@
+
+class AbstractAppendTest extends RenderingTest {
+
+  constructor() {
+    super();
+
+    this.components = [];
+    this.ids = [];
+  }
+
+  teardown() {
+    this.component = null;
+
+    this.components.forEach(component => {
+      this.runTask(() => component.destroy());
+    });
+
+    this.ids.forEach(id => {
+      let $element = jQuery(id).remove();
+      this.assert.strictEqual($element.length, 0, `Should not leak element: #${id}`);
+    });
+
+    super();
+  }
+
+  /* abstract append(component): Element; */
+
+  didAppend(component) {
+    this.components.push(component);
+    this.ids.push(component.elementId);
+  }
+
+  ['@test appending, updating and destroying a single component'](assert) {
+    let willDestroyCalled = 0;
+
+    this.registerComponent('x-parent', {
+      ComponentClass: Component.extend({
+        layoutName: 'components/x-parent',
+
+        willDestroyElement() {
+          willDestroyCalled++;
+        }
+      }),
+
+      template: '[parent: {{foo}}]{{#x-child bar=foo}}[yielded: {{foo}}]{{/x-child}}'
+    });
+
+    this.registerComponent('x-child', {
+      ComponentClass: Component.extend({
+        tagName: ''
+      }),
+
+      template: '[child: {{bar}}]{{yield}}'
+    });
+
+    let XParent = this.owner._lookupFactory('component:x-parent');
+
+    this.component = XParent.create({ foo: 'zomg' });
+
+    assert.ok(!this.component.element, 'precond - should not have an element');
+
+    this.element = this.append(this.component);
+
+    let componentElement = this.component.element;
+
+    this.assertComponentElement(componentElement, { content: '[parent: zomg][child: zomg][yielded: zomg]' });
+
+    assert.equal(componentElement.parentElement, this.element, 'It should be attached to the target');
+
+    this.runTask(() => this.rerender());
+
+    this.assertComponentElement(componentElement, { content: '[parent: zomg][child: zomg][yielded: zomg]' });
+
+    assert.equal(componentElement.parentElement, this.element, 'It should be attached to the target');
+
+    this.runTask(() => set(this.component, 'foo', 'wow'));
+
+    this.assertComponentElement(componentElement, { content: '[parent: wow][child: wow][yielded: wow]' });
+
+    assert.equal(componentElement.parentElement, this.element, 'It should be attached to the target');
+
+    this.runTask(() => set(this.component, 'foo', 'zomg'));
+
+    this.assertComponentElement(componentElement, { content: '[parent: zomg][child: zomg][yielded: zomg]' });
+
+    assert.equal(componentElement.parentElement, this.element, 'It should be attached to the target');
+
+    this.runTask(() => this.component.destroy());
+
+    assert.ok(!this.component.element, 'It should not have an element');
+
+    assert.ok(!componentElement.parentElement, 'The component element should be detached');
+
+    this.assert.equal(willDestroyCalled, 1);
+  }
+
+  ['@htmlbars appending, updating and destroying multiple components']() {
```

Comment:
Need to add assert argument to test function declaration since it is used below.